hail/src/main/scala/is/hail/io/bgen/BgenRDDPartitions.scala

```diff
@@ -15,158 +17,159 @@ trait BgenPartition extends Partition {
-        val nPartitions = fileNPartitions(fileIndex)
-        using(new OnDiskBTreeIndexToValue(file.path + ".idx", hConf)) { index =>
-          includedVariantsPerFile.get(file.path) match {
-            case None =>
-              val partNVariants = partition(file.nVariants, nPartitions)
-              val partFirstVariantIndex = partNVariants.scan(0)(_ + _).init
-              val partFirstByteOffset = index.positionOfVariants(partFirstVariantIndex)
-              var i = 0
-              while (i < nPartitions) {
-                partitions += BgenPartitionWithoutFilter(
-                  file.path,
-                  file.compressed,
-                  partFirstVariantIndex(i),
-                  partFirstByteOffset(i),
-                  if (i < nPartitions - 1) partFirstByteOffset(i + 1) else file.fileByteSize,
-                  partitions.length,
-                  sHadoopConfBc,
-                  settings
-                )
-                i += 1
-              }
-            case Some(variantIndices) =>
-              val partNVariants = partition(variantIndices.length, nPartitions)
-              val partFirstVariantIndex = partNVariants.scan(0)(_ + _).init
-              val variantIndexByteOffset = index.positionOfVariants(variantIndices.toArray)
-              var i = 0
-              while (i < nPartitions) {
-                var firstVariantIndex = partFirstVariantIndex(i)
-                var lastVariantIndex = firstVariantIndex + partNVariants(i)
-                partitions += BgenPartitionWithFilter(
-                  file.path,
-                  file.compressed,
-                  partitions.length,
-                  variantIndexByteOffset.slice(firstVariantIndex, lastVariantIndex),
-                  variantIndices.slice(firstVariantIndex, lastVariantIndex).toArray,
-                  sHadoopConfBc,
-                  settings
-                )
-                i += 1
-              }
+        using(IndexReader(hConf, file.indexPath)) { index =>
+          val nPartitions = fileNPartitions(fileIndex)
+          val partNVariants = partition(file.nVariants.toInt, nPartitions)
+          val partFirstVariantIndex = partNVariants.scan(0)(_ + _).init
+          var i = 0
+          while (i < nPartitions) {
+            val firstVariantIndex = partFirstVariantIndex(i)
+            val lastVariantIndex = firstVariantIndex + partNVariants(i)
+            val partitionIndex = partitions.length
+
+            partitions += LoadBgenPartition(
+              file.path,
+              file.indexPath,
+              filterPartition = null,
+              file.header.compressed,
+              file.skipInvalidLoci,
+              file.contigRecoding,
+              partitionIndex,
+              firstVariantIndex,
+              lastVariantIndex,
+              sHadoopConfBc
+            )
+
+            if (firstVariantIndex < file.nVariants && lastVariantIndex <= file.nVariants) {
```

Comment:
I think I added this because I was getting errors if the number of partitions specified was larger than the number of variants in the file. I could have made it such that we set the nPartitions to not be greater than the number of variants or we could allow empty partitions which was the existing behavior. I wasn't sure what the behavior of nPartitions should be. The code would be a lot simpler if we didn't have empty partitions. I think this is probably a better way to go.